worker/caasapplicationprovisioner/worker.go

```diff
@@ -163,20 +188,36 @@ func (p *provisioner) loop() error {
 				}
 
 				config := AppWorkerConfig{
-					Name:       app,
+					Name:       appName,
 					Facade:     p.facade,
 					Broker:     p.broker,
 					ModelTag:   p.modelTag,
 					Clock:      p.clock,
-					Logger:     p.logger.Child("applicationworker"),
+					Logger:     p.logger.Child(appName),
 					UnitFacade: p.unitFacade,
+					ShutDownCleanUpFunc: func() {
+						go p.shutDownAppWorker(appName)
```

Comment:
I don't think this needs to be a go routine

Reply:
I think it will be a deadlock if it's not running in a separate goroutine because StopAndRemoveWorker will block the `loop()` to exit.

Reply:
Discussed live, this approach breaks layering - a worker should not be responsible for removing itself from a runner create a layer above.
The correct approach is to use a defer to transform err nil to ErrApplicationDone and have the IsFatal() method on the runner return true for these errors. That causes the runner to not restart the worker and instead remove it.

Reply:
The change in appworker.Notify() fixed the blocking issue in the applicationprovisioner worker.
Combining with the applicationprovisioner level shutdown fix, the whole shutdown process should work as expected now.
So no need this ShutDownCleanUpFunc now.